### PR TITLE
Add Runtime.WindowsRuntime tests

### LINKED ISSUE #21303: Subclass of MemoryStream can cause a Debug.Assert to be hit in WindowsRuntimeBuffer
```cs
[Fact]
public void GetWindowsRuntimeBuffer_BufferWithLengthGreaterThanIntMax_Throws()
{
    var memoryStream = new SubMemoryStream(10);
    WindowsRuntimeBufferExtensions.GetWindowsRuntimeBuffer(memoryStream, 0, 0);
}

public class SubMemoryStream : MemoryStream
{
    public SubMemoryStream(int capacity) : base(capacity) { }

    public override long Length => long.MaxValue;
}
```

Stack:
```
FailFast:

     at System.Diagnostics.Debug.Assert(Boolean condition, String message, String detailMessage)
     at System.Diagnostics.Debug.Assert(Boolean condition)
     at System.Runtime.InteropServices.WindowsRuntime.WindowsRuntimeBufferExtensions.GetWindowsRuntimeBuffer(MemoryStre
  am underlyingStream, Int32 positionInStream, Int32 length) in C:\Users\hugh\Documents\GitHub\corefx\src\System.Runtim
  e.WindowsRuntime\src\System\Runtime\InteropServices\WindowsRuntime\WindowsRuntimeBufferExtensions.cs:line 445
     at System.Runtime.InteropServices.WindowsRuntime.Tests.WindowsRuntimeBufferExtensionsTests.GetWindowsRuntimeBuffer
  _BufferWithLengthGreaterThanIntMax_Throws() in C:\Users\hugh\Documents\GitHub\corefx\src\System.Runtime.WindowsRuntim
  e\tests\System\Runtime\InteropServices\WindowsRuntimeBufferExtensionsTests.cs:line 199
     at System.RuntimeMethodHandle.InvokeMethod(Object target, Object[] arguments, Signature sig, Boolean constructor)
     at System.Reflection.RuntimeMethodInfo.UnsafeInvokeInternal(Object obj, Object[] parameters, Object[] arguments)
     at System.Reflection.RuntimeMethodInfo.Invoke(Object obj, BindingFlags invokeAttr, Binder binder, Object[] paramet
  ers, CultureInfo culture)
     at System.Reflection.MethodBase.Invoke(Object obj, Object[] parameters)
```

### PR DESCRIPTION
**Pull request by hughbe**
Did these about a week ago but forgot to ever send it in

Fixes #21303

### CODE REVIEW

**Comment by hughbe** [comment]
@safern would you mind testing this under uapaot? thanks

**Comment by safern** [comment]
> @safern would you mind testing this under uapaot if applicable? thanks

Sure. Doing it now

**Comment by safern** [comment]
I've checked and it shows 18 failed tests. This is the log: 

```
Windows.Foundation.Tests.SizeTests.Ctor_NegativeHeight_ThrowsArgumentOutOfRangeException(height: -1) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentOutOfRangeException)
        Actual:   typeof(System.ArgumentException): Value does not fall within the expected range.
        Stack Trace:
           D:\repos\corefxCopy\corefx\src\System.Runtime.WindowsRuntime\tests\Windows\Foundation\SizeTests.cs(51,0): at Windows.Foundation.Tests.SizeTests.<>c__DisplayClass3_0.<Ctor_NegativeHeight_ThrowsArgumentOutOfRangeException>b__0()
     Windows.Foundation.Tests.SizeTests.Ctor_NegativeHeight_ThrowsArgumentOutOfRangeException(height: -ì) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentOutOfRangeException)
        Actual:   typeof(System.ArgumentException): Value does not fall within the expected range.
        Stack Trace:
           D:\repos\corefxCopy\corefx\src\System.Runtime.WindowsRuntime\tests\Windows\Foundation\SizeTests.cs(51,0): at Windows.Foundation.Tests.SizeTests.<>c__DisplayClass3_0.<Ctor_NegativeHeight_ThrowsArgumentOutOfRangeException>b__0()
     Windows.Foundation.Tests.SizeTests.Height_SetNegative_ThrowsArgumentOutOfRangeException(height: -1) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentOutOfRangeException)
        Actual:   typeof(System.ArgumentException): Value does not fall within the expected range.
        Stack Trace:
           D:\repos\corefxCopy\corefx\src\System.Runtime.WindowsRuntime\tests\Windows\Foundation\SizeTests.cs(91,0): at Windows.Foundation.Tests.SizeTests.<>c__DisplayClass7_0.<Height_SetNegative_ThrowsArgumentOutOfRangeException>b__0()
     Windows.Foundation.Tests.SizeTests.Height_SetNegative_ThrowsArgumentOutOfRangeException(height: -ì) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentOutOfRangeException)
        Actual:   typeof(System.ArgumentException): Value does not fall within the expected range.
        Stack Trace:
           D:\repos\corefxCopy\corefx\src\System.Runtime.WindowsRuntime\tests\Windows\Foundation\SizeTests.cs(91,0): at Windows.Foundation.Tests.SizeTests.<>c__DisplayClass7_0.<Height_SetNegative_ThrowsArgumentOutOfRangeException>b__0()
     Windows.Foundation.Tests.SizeTests.Ctor_NegativeWidth_ThrowsArgumentOutOfRangeException(width: -1) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentOutOfRangeException)
        Actual:   typeof(System.ArgumentException): Value does not fall within the expected range.
        Stack Trace:
           D:\repos\corefxCopy\corefx\src\System.Runtime.WindowsRuntime\tests\Windows\Foundation\SizeTests.cs(43,0): at Windows.Foundation.Tests.SizeTests.<>c__DisplayClass2_0.<Ctor_NegativeWidth_ThrowsArgumentOutOfRangeException>b__0()
     Windows.Foundation.Tests.SizeTests.Ctor_NegativeWidth_ThrowsArgumentOutOfRangeException(width: -ì) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentOutOfRangeException)
        Actual:   typeof(System.ArgumentException): Value does not fall within the expected range.
        Stack Trace:
           D:\repos\corefxCopy\corefx\src\System.Runtime.WindowsRuntime\tests\Windows\Foundation\SizeTests.cs(43,0): at Windows.Foundation.Tests.SizeTests.<>c__DisplayClass2_0.<Ctor_NegativeWidth_ThrowsArgumentOutOfRangeException>b__0()
     Windows.Foundation.Tests.SizeTests.Width_SetNegative_ThrowsArgumentOutOfRangeException(width: -1) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentOutOfRangeException)
        Actual:   typeof(System.ArgumentException): Value does not fall within the expected range.
        Stack Trace:
           D:\repos\corefxCopy\corefx\src\System.Runtime.WindowsRuntime\tests\Windows\Foundation\SizeTests.cs(71,0): at Windows.Foundation.Tests.SizeTests.<>c__DisplayClass5_0.<Width_SetNegative_ThrowsArgumentOutOfRangeException>b__0()
     Windows.Foundation.Tests.SizeTests.Width_SetNegative_ThrowsArgumentOutOfRangeException(width: -ì) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentOutOfRangeException)
        Actual:   typeof(System.ArgumentException): Value does not fall within the expected range.
        Stack Trace:
           D:\repos\corefxCopy\corefx\src\System.Runtime.WindowsRuntime\tests\Windows\Foundation\SizeTests.cs(71,0): at Windows.Foundation.Tests.SizeTests.<>c__DisplayClass5_0.<Width_SetNegative_ThrowsArgumentOutOfRangeException>b__0()
     Windows.Foundation.Tests.RectTests.Ctor_NegativeHeight_ThrowsArgumentOutOfRangeException(height: -1) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentOutOfRangeException)
        Actual:   typeof(System.ArgumentException): Value does not fall within the expected range.
        Stack Trace:
           D:\repos\corefxCopy\corefx\src\System.Runtime.WindowsRuntime\tests\Windows\Foundation\RectTests.cs(66,0): at Windows.Foundation.Tests.RectTests.<>c__DisplayClass3_0.<Ctor_NegativeHeight_ThrowsArgumentOutOfRangeException>b__0()
     Windows.Foundation.Tests.RectTests.Ctor_NegativeHeight_ThrowsArgumentOutOfRangeException(height: -ì) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentOutOfRangeException)
        Actual:   typeof(System.ArgumentException): Value does not fall within the expected range.
        Stack Trace:
           D:\repos\corefxCopy\corefx\src\System.Runtime.WindowsRuntime\tests\Windows\Foundation\RectTests.cs(66,0): at Windows.Foundation.Tests.RectTests.<>c__DisplayClass3_0.<Ctor_NegativeHeight_ThrowsArgumentOutOfRangeException>b__0()
     Windows.Foundation.Tests.RectTests.Height_SetNegative_ThrowsArgumentOutOfRangeException(height: -1) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentOutOfRangeException)
        Actual:   typeof(System.ArgumentException): Value does not fall within the expected range.
        Stack Trace:
           D:\repos\corefxCopy\corefx\src\System.Runtime.WindowsRuntime\tests\Windows\Foundation\RectTests.cs(188,0): at Windows.Foundation.Tests.RectTests.<>c__DisplayClass16_0.<Height_SetNegative_ThrowsArgumentOutOfRangeException>b__0()
     Windows.Foundation.Tests.RectTests.Height_SetNegative_ThrowsArgumentOutOfRangeException(height: -ì) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentOutOfRangeException)
        Actual:   typeof(System.ArgumentException): Value does not fall within the expected range.
        Stack Trace:
           D:\repos\corefxCopy\corefx\src\System.Runtime.WindowsRuntime\tests\Windows\Foundation\RectTests.cs(188,0): at Windows.Foundation.Tests.RectTests.<>c__DisplayClass16_0.<Height_SetNegative_ThrowsArgumentOutOfRangeException>b__0()
     Windows.Foundation.Tests.RectTests.Ctor_NegativeWidth_ThrowsArgumentOutOfRangeException(width: -1) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentOutOfRangeException)
        Actual:   typeof(System.ArgumentException): Value does not fall within the expected range.
        Stack Trace:
           D:\repos\corefxCopy\corefx\src\System.Runtime.WindowsRuntime\tests\Windows\Foundation\RectTests.cs(58,0): at Windows.Foundation.Tests.RectTests.<>c__DisplayClass2_0.<Ctor_NegativeWidth_ThrowsArgumentOutOfRangeException>b__0()
     Windows.Foundation.Tests.RectTests.Ctor_NegativeWidth_ThrowsArgumentOutOfRangeException(width: -ì) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentOutOfRangeException)
        Actual:   typeof(System.ArgumentException): Value does not fall within the expected range.
        Stack Trace:
           D:\repos\corefxCopy\corefx\src\System.Runtime.WindowsRuntime\tests\Windows\Foundation\RectTests.cs(58,0): at Windows.Foundation.Tests.RectTests.<>c__DisplayClass2_0.<Ctor_NegativeWidth_ThrowsArgumentOutOfRangeException>b__0()
     Windows.Foundation.Tests.RectTests.Width_SetNegative_ThrowsArgumentOutOfRangeException(width: -1) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentOutOfRangeException)
        Actual:   typeof(System.ArgumentException): Value does not fall within the expected range.
        Stack Trace:
           D:\repos\corefxCopy\corefx\src\System.Runtime.WindowsRuntime\tests\Windows\Foundation\RectTests.cs(171,0): at Windows.Foundation.Tests.RectTests.<>c__DisplayClass14_0.<Width_SetNegative_ThrowsArgumentOutOfRangeException>b__0()
     Windows.Foundation.Tests.RectTests.Width_SetNegative_ThrowsArgumentOutOfRangeException(width: -ì) [FAIL]
        Assert.Throws() Failure
        Expected: typeof(System.ArgumentOutOfRangeException)
        Actual:   typeof(System.ArgumentException): Value does not fall within the expected range.
        Stack Trace:
           D:\repos\corefxCopy\corefx\src\System.Runtime.WindowsRuntime\tests\Windows\Foundation\RectTests.cs(171,0): at Windows.Foundation.Tests.RectTests.<>c__DisplayClass14_0.<Width_SetNegative_ThrowsArgumentOutOfRangeException>b__0()
System.Runtime.InteropServices.WindowsRuntime.Tests.WindowsRuntimeBufferExtensionsTests.AsStream_Buffer_Success [FAIL]
        System.Reflection.MissingMetadataException : This operation cannot be carried out because metadata for the following object was removed for performance reasons:\n\n  EETypeRva:0x0008D448\n\nNo further information is available. Rebuild in debug mode for better information.\n\n
System.Runtime.InteropServices.WindowsRuntime.Tests.WindowsRuntimeBufferExtensionsTests.IsSameData_Buffer_ReturnsExpected [FAIL]
        System.Reflection.MissingMetadataException : This operation cannot be carried out because metadata for the following object was removed for performance reasons:\n\n  EETypeRva:0x0008D448\n\nNo further information is available. Rebuild in debug mode for better information.\n\n

Discovering: System.Runtime.WindowsRuntime.Tests
  Discovered:  System.Runtime.WindowsRuntime.Tests
  Starting:    System.Runtime.WindowsRuntime.Tests
  Finished:    System.Runtime.WindowsRuntime.Tests
  
  === TEST EXECUTION SUMMARY ===
D:\repos\corefxCopy\corefx\Tools\tests.targets(346,5): warning :    System.Runtime.WindowsRuntime.Tests  Total: 264, Errors: 0, Failed: 18, Skipped: 0, Time: 0.323s
```

**Comment by tijoytom** [comment]
@hughbe For the exception difference create a bug with UWP6.0 milestone and disable the test for UAPAOT against the bug.

You might need an RD.XML for the System.Reflection.MissingMetadataException  exception , Do you use any reflection in the test?

**Comment by hughbe** [comment]
The code for the failing metadata test is as follows. No reflection here - do you know why it's failing?

It could be xunit reflecting on the IBuffer? What do you think

```cs
public static IEnumerable<object[]> AsStream_TestData()
        {
            yield return new object[] { new byte[0].AsBuffer(), new byte[0] };
            yield return new object[] { new byte[] { 1, 2, 3 }.AsBuffer(), new byte[] { 1, 2, 3 } };
            yield return new object[] { new byte[] { 1, 2, 3 }.AsBuffer(1, 2), new byte[] { 2, 3 } };
        }

        [Theory]
        [MemberData(nameof(AsStream_TestData))]
        public void AsStream_Buffer_Success(IBuffer buffer, byte[] expected)
        {
            using (MemoryStream stream = (MemoryStream)buffer.AsStream())
            {
                Assert.Equal(expected.Length, stream.Length);
                Assert.True(stream.CanWrite);

                Assert.Equal(expected, stream.ToArray());
            }
        }
```

**Comment by safern** [comment]
I just built it without optimizations to get the real type that needs to be added to the rd.xml file. 

```
System.Runtime.InteropServices.WindowsRuntime.Tests.WindowsRuntimeBufferExtensionsTests.IsSameData_Buffer_ReturnsExpected [FAIL]
        System.Reflection.MissingMetadataException : 'System.Runtime.InteropServices.WindowsRuntime.WindowsRuntimeBuffer' is missing metadata.
```

Here is an example of the rd.xml file: 

https://github.com/dotnet/corefx/blob/master/src/System.Linq/tests/Resources/System.Linq.Tests.rd.xml

**Comment by yizhang82** [comment]
@hughbe @safern The exception doesn't make sense to me. It looks like there is a bug in latest .NET native. We should not require people to add WindowsRuntimeBuffer in rd.xml. @tijoytom Can you help take a look? 
The exception diff seems to be a result of overly aggressive optimization. I remember @hoyMS mentioned something like this in the past. If we keep running into issues like these, we should consider turning off that particular optimization, at least in our testing. And if that optimization doesn't bring us a lot of value, we should consider simply turn it off.

**Comment by hughbe** [comment]
I've updated the PR to fix the aot failures

File: src/System.Runtime.WindowsRuntime/tests/Resources/System.Runtime.WindowsRuntime.Tests.rd.xml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+<Directives xmlns="http://schemas.microsoft.com/netfx/2013/01/metadata">
+  <Library>
```
Comment (safern) [inline src/System.Runtime.WindowsRuntime/tests/Resources/System.Runtime.WindowsRuntime.Tests.rd.xml]: As per comment: https://github.com/dotnet/corefx/pull/21658#issuecomment-311869642

You might want to log an issue to investigate this and add it to this file.

**Comment by tijoytom** [comment]
@hughbe Thanks for adding the test . Looks good now ..

**Comment by yizhang82** [comment]
@tijoytom as we discussed earlier, please double check the exception diff and the requirement of rd.xml to see if they are real issues.

**Comment by yizhang82** [comment]
@hughbe Thanks a lot for your contribution! Really appreciate it!

**Pull request merged**